library/Core/src/Juvix/Core/IR/Types/Base.hs

```diff
@@ -276,5 +338,56 @@ type GlobalsWith ty ext primTy primVal =
 type RawGlobals' ext primTy primVal =
   GlobalsWith (Term' ext) ext primTy primVal
 
-type Globals' extV extT primTy primVal =
-  GlobalsWith (Value' extV) extT primTy primVal
+type Globals' ext primTy primVal =
+  GlobalsWith (Value' ext) ext primTy primVal
+
+type Telescope ext primTy primVal =
+  [(Name, Term' ext primTy primVal)]
+
+data FunClause' ext primTy primVal
+  = -- | Clause has been labelled as unreachable by the coverage checker.
+    --   @Nothing@ means coverage checker has not run yet (clause may be unreachable).
+    --   @Just False@ means clause is not unreachable.
+    --   @Just True@ means clause is unreachable.
+    FunClause
+      { -- | @Δ@: The types of the pattern variables in dependency order.
+        -- , namedClausePats :: NAPs (Using Name instead atm)
+        -- ^ @Δ ⊢ ps@.  The de Bruijn indices refer to @Δ@.
+        clauseTel :: Telescope ext primTy primVal,
+        namedClausePats :: [Pattern' ext primTy primVal], --TODO [SplitPattern]
+
+        -- | @Just v@ with @Δ ⊢ v@ for a regular clause, or @Nothing@ for an
+        --   absurd one.
+        clauseBody :: Term' ext primTy primVal,
+        -- | @Δ ⊢ t@.  The type of the rhs under @clauseTel@.
+        clauseType :: Maybe (Value' ext primTy primVal),
+        -- | Clause has been labelled as CATCHALL.
+        -- , clauseRecursive   :: Maybe Bool TODO add this when termination checking
+        -- ^ @clauseBody@ contains recursive calls; computed by termination checker.
+        --   @Nothing@ means that termination checker has not run yet,
+        --   or that @clauseBody@ contains meta-variables;
+        --   these could be filled with recursive calls later!
+        --   @Just False@ means definitely no recursive call.
+        --   @Just True@ means definitely a recursive call.
+        clauseCatchall :: Bool,
+        clauseUnreachable :: Maybe Bool
+      }
+  deriving (Generic)
+
+type Signature ty ext primTy primVal = Map.Map Name (SigDef ty ext primTy primVal)
+
+-- Return type of all type-checking functions.
+-- state monad for global signature
+type TypeCheck ty ext primTy primVal a = StateT (Signature ty ext primTy primVal) IO a
+
+data SigDef ty ext primTy primVal
```

Comment:
Merge with fromfrontend later.

Reply:
Andy and I agree that the one from fromFrondEnd is not quite the same and so we'll keep a separate one.